Added stubs for audio services.

St4rk opened the pull request:
I've stubbed the following audio services: audout:u(part was already stubbed), audin:u, audrec:u, audren:u, codecctl. I've done it to make the code review process easy with the next commits that will implement the audio.

lioncash commented on src/core/hle/service/audio/audin_u.h:
```diff
@@ -0,0 +1,24 @@
+
+#pragma once
+
+#include "core/hle/kernel/hle_ipc.h"
```
Correct me if I'm wrong, but wouldn't this include be able to be moved to the cpp file, or be turned into a forward declaration if necessary? (ditto for other headers)

St4rk replied:
You're probably right (about the hle_ipc.h). Initially I'd say that this is because the Kernel::HLERequestContext, but it doesn't matter anyway. The only reason that I was using it is because I used the other services implementation as a template for my implementation. If you want I can remove/move the "#include "core/hle/kernel/hle_ipc.h"" for all the files that I've stubbed, but... I guess the same should be done to every service as well (?)

St4rk replied:
(sorry if I said something really really wrong or stupid)

lioncash replied:
Nah, it's not stupid. I would suggest (re)moving them and using forward-decls if they truly aren't needed though, since it will marginally speed up compilation (especially if you consider how includes lead to other includes, leading to other includes, etc, etc. which is still stuff that needs to be discarded during compilation).

> but... I guess the same should be done to every service as well (?)

Yeah.

St4rk replied:
I've tried to build without the include and it builds fine, I'll start first removing from my files that I'll commit now and I'll check if I can do the same to the other services.

Thanks!

St4rk replied:
Oh, I guess the reason for it build without problem is because the service.h already includes the hle_ipc.h. However I guess we want to avoid any dependences from other header includes (service.h has the hle_ipc.h as I said), but if we remove it's probably going to break the services (assuming that we tried to remove the hle_ipc from the other services header). I'll think better how the file organization works and see what I can do (honestly this is driving me nuts lol).

lioncash replied:
You can just use a forward declaration instead of an include. i.e. in the header:

```cpp
namespace Kernel {
class HLERequestContext;
}
```

and just put the include in the cpp file to satisfy the dependency

St4rk replied:
neat, I'll try here, thank you very much for the clarification : )

lioncash replied:
no prob!

bunnei commented:
How much of this was actually "tested"? I suggest just stubbing functions that we actually can run on either homebrew or games, and verify that the stub is reasonably fine. The IPC can be finicky, just returning SUCCESS for everything (without properly constructing the response header or providing an interface) can result in things failing in a subtle way. See https://reswitched.github.io/SwIPC/ for a good reference on the correct responses for all of these functions.

Note, you can still add the decoding tables for these functions (and just set the function to nullptr), that way we'll get a ASSERT if it's hit and know what to implement :)

St4rk commented:
I've done what we've talked in Discord:
1. Changed the decoding tables for the functions that are going to be implemented, now they're nullptr that are going to trigger an assert and we know what we need to implement.
2. I've implemented AudOutU::ListAudioOuts (that was already started, but it was wrong), it might return a list of strings holding the name of the audio interfaces and the size of the list. As libtransistor example requires the list to have only one element, I'm returning only one, however, it's highly possible that it has more than one interface (maybe Docked ?) as Switch has support to surround sound(5.1 I guess?). Anyway I've left a TODO there.
3. Checked the  https://reswitched.github.io/SwIPC/ and I'll keep using it for reference as well.

For the next PR that I'll made will be the real implementation for audout:u and I hope the stubs make the code review for it more easy.

bunnei commented on src/core/CMakeLists.txt:
```diff
@@ -101,6 +101,16 @@ add_library(core STATIC
     hle/service/filesystem/filesystem.h
     hle/service/filesystem/fsp_srv.cpp
     hle/service/filesystem/fsp_srv.h
+    hle/service/audio/audin_u.cpp
```
Sorry I missed this earlier - this should be with the other audio includes above. Note, this list is in alphabetical order

bunnei commented on src/core/hle/service/audio/audrec_u.cpp:
```diff
@@ -0,0 +1,40 @@
+    }
+    ~IFinalOutputRecorder() = default;
+
+private:
```
Remove

bunnei commented on src/core/hle/service/audio/codecctl.h:
```diff
@@ -0,0 +1,25 @@
+
+class CodecCtl final : public ServiceFramework<CodecCtl> {
+public:
+    CodecCtl();
```
`explicit`

bunnei commented on src/core/hle/service/audio/audren_u.h:
```diff
@@ -0,0 +1,25 @@
+
+class AudRenU final : public ServiceFramework<AudRenU> {
+public:
+    AudRenU();
```
`explicit`

bunnei commented on src/core/hle/service/audio/audrec_u.h:
```diff
@@ -0,0 +1,25 @@
+
+class AudRecU final : public ServiceFramework<AudRecU> {
+public:
+    AudRecU();
```
`explicit`

St4rk commented on src/core/hle/service/audio/audout_u.cpp:
```diff
@@ -4,20 +4,58 @@
+};
+
 void AudOutU::ListAudioOuts(Kernel::HLERequestContext& ctx) {
     LOG_WARNING(Service, "(STUBBED) called");
```
¯\_(ツ)_/¯

bunnei commented:
👍 ty lgtm
